core/model/listeners.py

```diff
@@ -244,29 +188,96 @@ def last_update_time_change(target, value, oldvalue, initator):
     target.external_index_needs_updating()
 
 
-@event.listens_for(Equivalency, "before_delete")
-def equivalency_coverage_reset_on_equivalency_delete(mapper, _db, target: Equivalency):
+def equivalency_coverage_reset_on_equivalency_delete(
+    session: Session, target: Equivalency
+) -> None:
     """On equivalency delete reset the coverage records of ANY ids touching
     the deleted identifiers
-    TODO: This is a deprecated feature of listeners, we cannot write to the DB anymore
-    However we are doing this until we have a solution, ala queues
     """
-
-    session = Session(bind=_db)
     EquivalencyCoverageQueries.add_coverage_for_identifiers_chain(
         [target.input, target.output], _db=session
     )
 
 
-@event.listens_for(Identifier, "after_insert")
-def recursive_equivalence_on_identifier_create(mapper, connection, target: Identifier):
+def recursive_equivalence_on_identifier_create(
+    session: Session, instance: Identifier
+) -> None:
     """Whenever an Identifier is created we must atleast have the 'self, self'
     recursion available in the Recursives table else the queries will be incorrect"""
-    session = Session(bind=connection)
     session.add(
-        RecursiveEquivalencyCache(
-            parent_identifier_id=target.id, identifier_id=target.id
-        )
+        RecursiveEquivalencyCache(parent_identifier=instance, identifier=instance)
     )
-    session.commit()
-    session.close()
+
+
+def before_flush_event_listener(
```

Comment:
Does it make sense to create our own decorator `@before_flush(<model>, <action>) ` which does the type checking so we can modularize the implementation of this method.

Reply:
In case the commit gets rolled back after the flush I assume the listeners changes will also be rolled back?

Reply:
It took wayyyy longer then I expected, because I ran into nightmarish circular imports, but I added our own decorator here.